### PR TITLE
feature_filter in addition to feature_re, support for explain_prediction

### PR DESCRIPTION
**Pull request by lopuhin**
~~This is very much WIP, also missing sklearn tree ensemble explain_prediction support.~~

This branch brings two features:
- support ``feature_filter`` (a function) along with ``feature_re``, so now it's possible to exclude certain features
- support feature filtering for all explain_prediction functions

Also ``eli5.utils.mask`` now handles sparse "vectors".

### CODE REVIEW

File: eli5/_feature_names.py [hunk 1]
```diff
@@ -110,6 +117,20 @@ def filtered_by_re(self, feature_re):
             ),
             indices)
 
+    def handle_filter(self, feature_filter, feature_re, x=None):
+        if feature_re is not None and feature_filter:
+            raise TypeError('pass either feature_filter or feature_re')
```
Comment (kmike) [inline eli5/_feature_names.py]: I think ValueError is more common for such cases (see e.g. http://stackoverflow.com/questions/256222/which-exception-should-i-raise-on-bad-illegal-argument-combinations-in-python?rq=1)

**Comment by kmike** [comment]
Sorry, it seems I introduced a merge conflict in https://github.com/TeamHG-Memex/eli5/pull/133 :(

File: eli5/_feature_names.py (hunk 1 above)
Comment (lopuhin) [inline eli5/_feature_names.py]: Aha, thanks for the catch! I chose TypeError cause it's raised for missing/extra arguments, but ValueError really makes more sense, thanks for the link)

**Comment by lopuhin** [comment]
>  Sorry, it seems I introduced a merge conflict in #133 :(

np, that was hardly avoidable :)

**Comment by lopuhin** [comment]
@kmike this is finally ready for review - I implemented feature filtering in explain_prediction for sklearn trees and ensembles, and added more tests.

File: eli5/sklearn/explain_prediction.py [hunk 2]
```diff
@@ -107,7 +109,9 @@ def explain_prediction_linear_classifier(clf, doc,
                                          target_names=None,
                                          targets=None,
                                          feature_names=None,
-                                         vectorized=False
+                                         feature_re=None,
+                                         feature_filter=None,
```
Comment (kmike) [inline eli5/sklearn/explain_prediction.py]: Could you please add feature_filter to the docstring of this function?

File: eli5/sklearn/explain_prediction.py [hunk 3]
```diff
@@ -194,6 +197,8 @@ def explain_prediction_linear_regressor(reg, doc,
                                         target_names=None,
                                         targets=None,
                                         feature_names=None,
+                                        feature_re=None,
+                                        feature_filter=None,
```
Comment (kmike) [inline eli5/sklearn/explain_prediction.py]: Could you please add these arguments to the docstring?

File: eli5/sklearn/explain_weights.py [hunk 4]
```diff
@@ -155,7 +155,8 @@ def explain_linear_classifier_weights(clf,
                                       targets=None,
                                       feature_names=None,
                                       coef_scale=None,
-                                      feature_re=None
+                                      feature_re=None,
+                                      feature_filter=None,
```
Comment (kmike) [inline eli5/sklearn/explain_weights.py]: Here feature_filter is also not in the docstring

File: eli5/xgboost.py [hunk 5]
```diff
@@ -97,7 +99,10 @@ def explain_prediction_xgboost(
         target_names=None,
         targets=None,
         feature_names=None,
-        vectorized=False):
+        vectorized=False,
+        feature_re=None,
+        feature_filter=None,
```
Comment (kmike) [inline eli5/xgboost.py]: I don't mind any argument order, but it'd be nice to have it consistent - e.g. in `explain_prediction_tree_regressor` `vectorized` is the last argument.

File: tests/test_sklearn_explain_prediction.py [hunk 6]
```diff
@@ -303,3 +307,33 @@ def test_explain_tree_classifier_text(clf, newsgroups_train_big):
     res = explain_prediction(clf, docs[0], vec=vec, target_names=target_names)
     check_targets_scores(res)
     format_as_all(res, clf)
+
+
+def _assert_tree_explain_prediction_single_target(clf, X, feature_names):
```
Comment (kmike) [inline tests/test_sklearn_explain_prediction.py]: What do you think about moving this function to the top of the file, closer to other assert functions?

File: tests/test_sklearn_explain_weights.py [hunk 7]
```diff
@@ -270,15 +270,18 @@ def test_explain_linear_feature_re(newsgroups_train, vec):
         vec = InvertableHashingVectorizer(vec)
         vec.fit(docs)
 
-    res = explain_weights(clf, vec=vec, feature_re='^ath')
+    res = explain_weights(
+        clf, vec=vec, feature_re='^ath')
```
Comment (kmike) [inline tests/test_sklearn_explain_weights.py]: why wrap this line? :)

**Comment by kmike** [comment]
Looks great, thanks @lopuhin!

File: tests/test_sklearn_explain_weights.py (hunk 7 above)
Comment (lopuhin) [inline tests/test_sklearn_explain_weights.py]: Hah, that line suffered through some changes :) Thank for the catch! Fixed in 724c63a

File: tests/test_sklearn_explain_prediction.py (hunk 6 above)
Comment (lopuhin) [inline tests/test_sklearn_explain_prediction.py]: Yep, a good call, thanks! Fixed in 724c63a

File: eli5/xgboost.py (hunk 5 above)
Comment (lopuhin) [inline eli5/xgboost.py]: Right, I like that feature_re and feature_filter come after feature_names :) Fixed it here too in 724c63a

**Comment by lopuhin** [comment]
Thanks for review @kmike ! I added feature_name to docstrings (linking to explain_weights or explain_predictions docs, as for other common parameters), and fixed other issues.

File: eli5/sklearn/explain_prediction.py [hunk 8]
```diff
@@ -293,9 +297,26 @@ def explain_prediction_tree_classifier(
...
+    ``vec`` is a vectorizer instance used to transform
+    raw features to the input of the classifier ``clf``
+    (e.g. a fitted CountVectorizer instance); you can pass it
+    instead of ``feature_names``.
+
+    ``vectorized`` is a flag which tells eli5 if ``doc`` should be
+    passed through ``vec`` or not. By default it is False, meaning that
+    if ``vec`` is not None, ``vec.transform([doc])`` is passed to the
+    classifier. Set it to False if you're using ``vec`` to get feature names,
+    but ``doc`` is already vectorized.
```
Comment (kmike) [inline eli5/sklearn/explain_prediction.py]: See https://github.com/TeamHG-Memex/eli5/pull/134#discussion_r95250606 :P

File: eli5/sklearn/explain_prediction.py [hunk 9]
```diff
@@ -371,9 +398,26 @@ def explain_prediction_tree_regressor(
...
```
Comment (kmike) [inline eli5/sklearn/explain_prediction.py]: 😛

**Comment by lopuhin** [comment]
Oh, looks like I'm still missing some docstrings, let me double-check

File: eli5/sklearn/explain_prediction.py (hunk 8 above)
Comment (lopuhin) [inline eli5/sklearn/explain_prediction.py]: haha, right! :) Fixed in ea0c16a

**Comment by lopuhin** [comment]
> Oh, looks like I'm still missing some docstrings, let me double-check

Yeah, I missed some, I think now all feature_filter docstrings are in place, hopefully.

File: eli5/sklearn/explain_prediction.py [hunk 10]
```diff
@@ -371,9 +398,26 @@ def explain_prediction_tree_regressor(
...
+        feature_filter=None,
         vectorized=False):
     """ Explain prediction of a tree regressor.
 
+    See :func:`eli5.explain_prediction` for description of
+    ``top``, ``target_names``, ``targets``, ``feature_names``,
+    ``feature_re`` and ``feature_filter`` parameters.
+
+    ``vec`` is a vectorizer instance used to transform
+    raw features to the input of the classifier ``clf``
```
Comment (kmike) [inline eli5/sklearn/explain_prediction.py]: there is no ``clf`` argument

File: eli5/sklearn_crfsuite/explain_weights.py [hunk 11]
```diff
@@ -17,15 +17,21 @@ def explain_weights_sklearn_crfsuite(crf,
...
                                      target_names=None,
                                      targets=None,
-                                     feature_re=None):
-    """ Explain sklearn_crfsuite.CRF weights """
+                                     feature_re=None,
+                                     feature_filter=None):
+    """ Explain sklearn_crfsuite.CRF weights.
+
+    See :func:`eli5.explain_weights` for description of
+    ``top``, ``feature_names``,
```
Comment (kmike) [inline eli5/sklearn_crfsuite/explain_weights.py]: `target_names` and `targets` are also documented there, and there is no `feature_names` argument for explain_weights_sklearn_crfsuite

File: eli5/xgboost.py [hunk 12]
```diff
@@ -97,10 +105,28 @@ def explain_prediction_xgboost(
...
+        ):
     """ Return an explanation of XGBoost prediction (via scikit-learn wrapper
     XGBClassifier or XGBRegressor) as feature weights.
 
+    See :func:`eli5.explain_prediction` for description of
+    ``top``, ``target_names``, ``targets``, ``feature_names``,
+    ``feature_re`` and ``feature_filter`` parameters.
+
+    ``vec`` is a vectorizer instance used to transform
+    raw features to the input of the classifier ``clf``
```
Comment (kmike) [inline eli5/xgboost.py]: there is no ``clf`` argument

File: eli5/sklearn/explain_prediction.py (hunk 10 above)
Comment (lopuhin) [inline eli5/sklearn/explain_prediction.py]: Must be some GH bug, I bet the comment was for the correct line (this https://github.com/TeamHG-Memex/eli5/pull/132/files/ea0c16abdd1629890ed97dd352d623c20e33de98#diff-5a5dfe2e408016505008ee23671cb61fR411), but now it's show in the wrong place for me.
Fixed in dec5f20. Thanks for taking care of the docs, I didn't check them careful enough.
